### ISSUE TITLE
How to properly configure for ASP.NET Core 2 and Azure App Service

### ISSUE THREAD

**Issue by fgauna12** [serilog/serilog-sinks-applicationinsights#37]
I've had difficulty figuring out how to configure a Serilog App Insights sink with ASP.NET core 2.

For Azure app services, the instrumentation key is loaded from `APPINSIGHTS_INSTRUMENTATIONKEY` config setting.

I would like to keep using that config setting instead of recreating it. 

So after following the instructions from the main Serilog repository, I ended up with this:

``` csharp
var loggerConfiguration = new LoggerConfiguration()
    .MinimumLevel.Debug()
    .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
    .Enrich.FromLogContext()
    .WriteTo.Console();

//Create a builder just so that we can read the app insights instrumentation key from the config
var configuration = new ConfigurationBuilder()
    .SetBasePath(Directory.GetCurrentDirectory())
    .AddJsonFile("appsettings.json", optional: false)
    .AddJsonFile("appsettings.Development.json", optional: true)
    .AddJsonFile("appsettings.Staging.json", optional: true)
    .AddJsonFile("appsettings.Production.json", optional: true)
    .Build();

var appInsightsKey = configuration["APPINSIGHTS_INSTRUMENTATIONKEY"];
if (appInsightsKey != null)
{
    loggerConfiguration.WriteTo.ApplicationInsightsTraces(appInsightsKey);
}

Log.Logger = loggerConfiguration.CreateLogger();

try
{
    Log.Information("Starting web host");
    BuildWebHost(args).Run();

    return 0;
}
catch (Exception ex)
{
    Log.Fatal(ex, "Host terminated unexpectedly");

    return 1;
}
finally
{
    Log.CloseAndFlush();
}
```

Is there a better way to do this?

**Comment by robdmoore** [comment]
App insights automatically picks up that setting ambiently if it exists.

Just don't pass in an instrumentation key. From memory you might need to pass in a TelemetryClient instead, just new one up without passing constructor params.

--

**Comment by nblumhardt** [comment]
Just in case you do wish to create a configuration as you're doing, there's a slightly more succinct way to do it without specifying each individual environment name:

https://github.com/serilog/serilog-aspnetcore/blob/dev/samples/SimpleWebSample/Program.cs#L16

Hope this helps, let us know how you go! :-)

**Comment by fgauna12** [comment]
Thanks for the suggestions!

I tried passing in an empty `TelemetryClient` but it's not working for me. I don't see any data on App Insights.

``` csharp

var loggerConfiguration = new LoggerConfiguration()
                .MinimumLevel.Debug()
                .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
                .Enrich.FromLogContext()
                .WriteTo.Console()
                .WriteTo.ApplicationInsightsTraces(new TelemetryClient());

```

If there aren't any more suggestions, I'll go with the more succinct version of what I had.

**Comment by robdmoore** [comment]
Hmm, looking at https://github.com/Microsoft/ApplicationInsights-aspnetcore/wiki/Configure it seems the .net core one uses ApplicationInsights:InstrumentationKey rather than APPINSIGHTS_INSTRUMENTATIONKEY. That’s disappointing.

Your options would probably be:

1. Set the ApplicationInsights:InstrumentationKey setting within you web app to the instrumentation key (you can do that using ARM if that’s how you provision your site infrastructure. Then you should be able to use new TelemetryClient()
2. Use your technique above
3. Set the ApplicationInsights:InstrumentationKey setting in code, from the APPINSIGHTS_INSTRUMENTATIONKEY setting and then use new TelemetryClient()

**Comment by jmingli** [comment]
@fgauna12 

I am having the same issue, have you figured out the solution yet?

**Comment by fgauna12** [comment]
@jmingli I stuck with my original solution and the suggestions from @nblumhardt 

``` csharp

var loggerConfiguration = new LoggerConfiguration()
    .MinimumLevel.Debug()
    .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
    .Enrich.FromLogContext()
    .WriteTo.Console();

//Create a builder just so that we can read the app insights instrumentation key from the config
var configuration = new ConfigurationBuilder()
    .SetBasePath(Directory.GetCurrentDirectory())
    .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
    .AddJsonFile($"appsettings.{Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ?? "Production"}.json", optional: true)
    .Build();

var appInsightsKey = configuration["APPINSIGHTS_INSTRUMENTATIONKEY"];
if (appInsightsKey != null)
{
    loggerConfiguration.WriteTo.ApplicationInsightsTraces(appInsightsKey);
}

Log.Logger = loggerConfiguration.CreateLogger();

try
{
    Log.Information("Starting web host");
    BuildWebHost(args).Run();

    return 0;
}
catch (Exception ex)
{
    Log.Fatal(ex, "Host terminated unexpectedly");

    return 1;
}
finally
{
    Log.CloseAndFlush();
}

```

**Comment by Triwaters** [comment]
@nblumhardt @fgauna12 

I may be confused but have to ask... is newing up a Configuration in Program.cs going to run afoul of the discouraged activities noted here? http://thedatafarm.com/dotnet/twitter-education-re-aspnet-core-scope/  These may be utterly unrelated topics, which is why I am asking. Maybe so.

If the topics *are* related, I am further wondering if what we're talking about here is forgivable because (I'm guessing?) the thing being scoped is a small object that's not consuming (many) resources or holding open a connection?

Apologies if this is a rambling question. I'm not fluent in this stuff yet.

**Comment by nblumhardt** [comment]
@Triwaters good question! No, this is all happening before IoC is bootstrapped because it _needs to_, if you want to capture exceptions thrown during IoC configuration in your application logs.

The default templates don't work this way, but there are differing perspectives - some want to configure logging within IoC because they want to inject bits and pieces into their logging pipeline through the container.

I started the Autofac project back in 2007, and since then I've seen so many application startup failures occur during container initialization that I have the stack traces and error messages permanently burned into my brain. I'd never configure logging after IoC, personally - it sounds like a perfect recipe for blundering around in the dark trying to figure out why your app won't start up in production :-)

**Comment by Triwaters** [comment]
@nblumhardt you're totally making sense, and in my short tenure here in .NET land I've already got a few scorch marks on my hands from precisely that problem.

One question, to make sure I understand you. I'm thinking you're saying... We still set up configuration in Startup.cs -- for general use via DI from that point on. But prior to that in Program.cs we load up a configuration object for the (possibly sole) purpose of setting up logging properly right out of the gate. We know that the configuration object will persist, unused, for the duration... but we also know also that it's small, not hurting anything, and we don't care... because it's worth it just to get logging going at the very, very start. Am I understanding you? If so, I'm very excited to have the freedom to use that Configuration object for a few other things in Program.cs!

Thank you so much for your help on this. I've been devouring logging information over the past two days, so I very much appreciated [this post of yours](https://nblumhardt.com/2017/08/use-serilog/) and was inspired by it to get Application Insights going properly in my project... thus ending up here.

And seeing as yesterday I discovered your Seq product (very nice) I owe you not one but two thanks and a free, unsolicited, endorsement...

**Comment by Triwaters** [comment]
Or are you suggesting that we dispense with Configuration work in Startup.cs entirely? I just noticed (late to the party) that ck me that your configuration is a public static class. The following code works anywhere in the project...

`var x = Program.Configuration["TestSetting"];`

If one should know the rules well before breaking them, I'm thinking I should be chary of this kind of rebellion for now! Seems we'd be breaking the recommended mold on multiple levels!

**Comment by Triwaters** [comment]
@nblumhardt Not to beat a dead horse (oh, wait, that's what I'm doing) but my last thought before retiring last night was... "What does this line of code accomplish?"

`.UseConfiguration(Configuration)`

After all, the preceding line...

`WebHost.CreateDefaultBuilder(args) `

already reads the same configuration data in by default, albeit with slightly different parameter settings. So the configuration data is already loaded. Am I missing something?

The above is not terribly important, since the end result is the same, but I'm still very curious to understand if you're proposing we use the public static Program.Configuration in the rest of the app? Or Configuration via std injection? And if the latter, can't/shouldn't the former be a private member, to make it more explicit that the Configuration in Program.cs is strictly used for logging startup?

**Comment by nblumhardt** [comment]
@Triwaters hi - yes, that's the proposal; here's a complete example: https://github.com/serilog/serilog-aspnetcore/blob/dev/samples/SimpleWebSample/Program.cs

I think a lot of questions can be answered by simply searching the solution-space for an alternative; there's really not any other obvious way to get logging hooked up before the rest of ASP.NET Core without juggling configuration around.

It'd be great if the ASP.NET Core and EF Core teams simplify and improve the logging-before-`WebHost` scenario in the future. Fingers crossed it will get on the radar eventually. Until then, this setup works well.

HTH,
Nick

**Comment by Triwaters** [comment]
Regarding configuration, I share your frustration. I spent a fair bit of time making trying to make sense (best I could) of it -- the 2.0 documentation was not always accurate! [(Witness my laments here, as Brian Wells/ Wellspring](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?tabs=visual-studio))

Having read all those docs, I can say with at least some confidence that this code in your sample...

`
        public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                   .UseStartup<Startup>()
                   .UseConfiguration(Configuration)
                   .UseSerilog()
                   .Build();
`

Can be reduced to this code...

`
        public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                   .UseStartup<Startup>()
                   .UseSerilog()
                   .Build();
`

without loss.

If I had time to prove it, I'd do so and then submit a PR on that repo. but not having proof in hand I just throw the suggestion out for consideration.

Thanks again!

I agree

**Comment by Triwaters** [comment]
@nblumhardt @fgauna12 I spent most of today trying to figure out why my log entries were not showing up in Application Insights. I think I found the issue, and am wondering if it relates at all to your original issues, @fgauna12 .

The problem in my code boiled down to the fact that the provided link...

[https://github.com/serilog/serilog-aspnetcore/blob/dev/samples/SimpleWebSample/Program.cs#L16](https://github.com/serilog/serilog-aspnetcore/blob/dev/samples/SimpleWebSample/Program.cs#L16)

Is missing the call to AddEnvironmentVariables() that is present in the original CreateDefaultBuilder()

That is a huge miss for my projects, because I use azure environment variables to point to different AI subscriptions. So because there is no read through via AddEnvironmentVariables() there are none of the ordinary overrides I have come to depend upon. HOWEVER (!) the old AppSettings.json file that gets deployed happily continues to supply the Dev AI subscription key, regardless of whether or not I'm in Staging or Prod... which inevitably I am by the time I get to azure web services.

In short, I think it's very important for users such as myself that AddEnvironmentVariables() be added to that sample repo. And since (I think??) my pattern of using Secrets in local Dev and environment variables in Azure (to populate private keys, etc.), I'm going to go out on a limb and say that I'm not off the beaten path, here.

Thoughts?

**Comment by nblumhardt** [comment]
Done! Thanks for the note.

**Comment by Triwaters** [comment]
Thanks! It was like the scene involving the pile of top-hats gathering in the woods in that movie, The Prestige. I spent the latter half of yesterday looking for comments in my Staging AI subscription, but they were just piling up in the Dev one the whole time, I'm pretty sure. :grinning:

**Comment by Triwaters** [comment]
@nblumhardt I'm very interested in your thoughts on whether it matters that the call to

`            WebHost.CreateDefaultBuilder(args)`

leads to this code being executed:
`
    .ConfigureLogging((hostingContext, logging) =>
      {
          logging.AddConfiguration(hostingContext.Configuration.GetSection("Logging"));
          logging.AddConsole();
          logging.AddDebug();
      })
`
I'm not keen to *not* call CreateDefaultBuilder, as I'd rather lean on the built-in defaults. But since logging calls are there, I am wondering if they produce multiple logging systems, as you discussed with David Fowler.

Thoughts?

**Comment by Triwaters** [comment]
For what it's worth, maybe of value to the next guy who wants to store his InstrumentationKey in a secrets file... this was the last leg in my journey. @nblumhardt I don't presume to think that this is how you'd do it. Looks like you'd prefer to deviate from the Microsoft course on Configuration DI, but for the sake of the steps required for acquisition of settings from secrets in local Dev and from environment settings in Azure, here's what I found worked for me.

```

    public class Program
    {
        public static int Main(string[] args)
        {            
            var configurationBuilder = new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddJsonFile($"appsettings.{Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") ?? "Production"}.json", optional: true)
                .AddEnvironmentVariables();
            if (Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == EnvironmentName.Development)
            {
                configurationBuilder.AddUserSecrets<Program>(false);
            }
            IConfiguration _configuration = configurationBuilder.Build();

            string _outputTemplate = "[{Timestamp:HH:mm:ss} {Level}] {SourceContext}{NewLine}{Message:lj}{NewLine}{Exception}{properties}{NewLine}";
            
            /*
             *  get the basics. For minimum log level (w/ overrides) in various environments, json config file seems
             *  the simplest approach, giving flexibility with tiers and also edits without recompile.
            */
            var loggerConfiguration = new LoggerConfiguration()
                .ReadFrom.Configuration(_configuration);

            // add the environment-specific nuance(s) that depend upon variables maintained elsewhere...
            if (Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == EnvironmentName.Development)
            {
                loggerConfiguration.WriteTo.Seq("http://localhost:" + Constants.DevWebsiteSeqPort.ToString());
            }

            // finish it off with settings unlikely to be changed across environments
            // and/or secrets best not kept in json repo files
            Log.Logger = loggerConfiguration
                .WriteTo.ApplicationInsightsEvents(_configuration["ApplicationInsights:InstrumentationKey"])
                .Enrich.FromLogContext()
                .WriteTo.Console(outputTemplate: _outputTemplate, theme: AnsiConsoleTheme.Literate)
                .WriteTo.File(
                      path: "[xxxxx]-.txt"
                    , restrictedToMinimumLevel: LogEventLevel.Information
                    , outputTemplate: _outputTemplate
                    , rollingInterval: RollingInterval.Day
                )
                .CreateLogger();
```

**Comment by joseclament** [comment]
I'm struggling to get my logs to application insight and can not get the logger started from the configuration file: @Triwaters do you have any idea?

Please see the below configuration in json:

    "Serilog": {
    "MinimumLevel": {
      "Default": "Verbose",
      "Override": {
        "Microsoft": "Warning",
        "System": "Warning"
      }
    },
    "WriteTo": [
      {
        "Name": "ApplicationInsightsEvents",
        "Args": {
          "InstrumentationKey": "xxxx-xxx-xxx-xxx-xxxx",
          "outputTemplate": "[{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} {Level} CorrelationId:{CorrelationId}][{SourceContext}]- {Message}{NewLine}{Exception}"
        }
      }
    ],
    "Properties": {
      "Application": "MyTestApplication",
      "Environment": "Production"
    }
  },

My Startup.cs have the below:

      public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
                .AddEnvironmentVariables();
            Configuration = builder.Build();
        }
 
    public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory, IOptions<ApplicationConfigurations> applicationConfigurations)
            {
      if (env.IsProduction())
                {
                    seriLogger = new LoggerConfiguration()
                    .Enrich.FromLogContext()
                    .ReadFrom.ConfigurationSection(Configuration.GetSection("Serilog"))
                    .CreateLogger();              
                }
                loggerFactory.AddSerilog(seriLogger);
                seriLogger.Information("Logger started");
                seriLogger.Information("Enviornment name :{EnvironmentName}", env.EnvironmentName);
    }

and finally the Program.cs 

     public class Program
    {
        public static void Main(string[] args)
        {
            var host = new WebHostBuilder()
                .UseKestrel()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .UseApplicationInsights()
                .Build();
            host.Run();
        }
    }

**Comment by Triwaters** [comment]
I'm a novice at this myself, and not at all sure. However I am suspecting your issue may related to you not having anything (see my posted code) like this line...

`
.WriteTo.ApplicationInsightsEvents(_configuration["ApplicationInsights:InstrumentationKey"])
`

Best of luck!
